tower-timeout/src/lib.rs

```diff
@@ -83,50 +98,114 @@ where T: Future,
         match self.response.poll() {
             Ok(Async::Ready(v)) => return Ok(Async::Ready(v)),
             Ok(Async::NotReady) => {}
-            Err(e) => return Err(Error::Inner(e)),
+            Err(e) => return Err(Error(Kind::Inner(e))),
         }
 
         // Now check the sleep
         match self.sleep.poll() {
             Ok(Async::NotReady) => Ok(Async::NotReady),
-            Ok(Async::Ready(_)) => Err(Error::Timeout),
-            Err(_) => Err(Error::Timeout),
+            Ok(Async::Ready(_)) => Err(Error(Kind::Elapsed)),
+            Err(_) => Err(Error(Kind::Elapsed)),
         }
     }
 }
 
-
 // ===== impl Error =====
 
 impl<T> fmt::Display for Error<T>
 where
     T: fmt::Display,
 {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        match *self {
-            Error::Inner(ref why) => fmt::Display::fmt(why, f),
-            Error::Timeout => f.pad("request timed out"),
+        match self.0 {
+            Kind::Inner(ref why) => fmt::Display::fmt(why, f),
+            Kind::Elapsed => f.pad("request timed out"),
+            Kind::Timer(ref why) => fmt::Display::fmt(why, f),
         }
     }
 }
 
 impl<T> error::Error for Error<T>
 where
-    T: error::Error,
+    T: error::Error + 'static,
 {
-    fn cause(&self) -> Option<&error::Error> {
-        if let Error::Inner(ref why) = *self {
+    fn description(&self) -> &str {
+        match self.0 {
+            Kind::Inner(_) => "inner service error",
+            Kind::Elapsed => "request timed out",
+            Kind::Timer(_) => "internal timer error"
+        }
+    }
+
+    fn source(&self) -> Option<&(error::Error + 'static)> {
+        let kind = &self.0;
+        if let Kind::Inner(ref why) = kind {
             Some(why)
         } else {
             None
         }
     }
+}
 
-    fn description(&self) -> &str {
-        match *self {
-            Error::Inner(_) => "inner service error",
-            Error::Timeout => "request timed out",
+// ===== impl Error =====
+
+impl<T> Error<T> {
+    /// Create a new `Error` representing the inner value completing with `Err`.
+    pub fn inner(err: T) -> Error<T> {
```

Comment:
TIOLI: I'm not sure if it's necessary for these constructors to be `pub`?

Reply:
I do see a potential use for being able to construct these in test code, though... 🤷‍♀️

Reply:
It's fine for it to pub for testing.

Reply:
Okay, let's make it pub, then!

Reply:
Should this be guarded by `#[cfg(test)]` then?

Reply:
When *depending* on a crate, even in tests, the dependency will not have been compiled with `--cfg test`. So setting `#[cfg(test)]` would only make it usable in unit tests within `tower-timeout`.

Reply:
@olix0r that would only be enabled when running lib tests in the crate proper.

If it is useful for downstream dependencies' tests, then it is part of the public API.